File: server/src/models/Stop.ts
import mongoose, { Document, Schema } from 'mongoose';
import { ObjectId, GeoPoint, Facilities } from './types';

// 정류장 인터페이스
export interface IStop extends Document {
  name: string;
  location: GeoPoint;
  routes: ObjectId[];
  facilities?: Facilities;
  description?: string;
  address?: string;
  imageUrl?: string;
  metadata?: Map<string, any>;
  createdAt: Date;
  updatedAt: Date;
}

// 정류장 스키마 정의
const stopSchema = new Schema<IStop>(
  {
    name: {
      type: String,
      required: [true, '정류장 이름은 필수입니다'],
      trim: true,
    },
    location: {
      type: {
        type: String,
        enum: ['Point'],
        default: 'Point',
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: [true, '정류장 위치는 필수입니다'],
      },
    },
    routes: [
      {
        type: Schema.Types.ObjectId,
        ref: 'Route',
      },
    ],
    facilities: {
      hasShelter: { type: Boolean, default: false },
      hasBench: { type: Boolean, default: false },
      hasLighting: { type: Boolean, default: false },
      hasElectronicDisplay: { type: Boolean, default: false },
      isAccessible: { type: Boolean, default: false },
    },
    description: {
      type: String,
      trim: true,
    },
    address: {
      type: String,
      trim: true,
    },
    imageUrl: {
      type: String,
      trim: true,
    },
    metadata: {
      type: Map,
      of: Schema.Types.Mixed,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// 지리적 인덱스 추가
stopSchema.index({ location: '2dsphere' });
stopSchema.index({ name: 'text' }); // 텍스트 검색을 위한 인덱스

// 가상 필드: 위도/경도
stopSchema.virtual('latitude').get(function (this: IStop) {
  return this.location?.coordinates?.[1];
});

stopSchema.virtual('longitude').get(function (this: IStop) {
  return this.location?.coordinates?.[0];
});

export default mongoose.model<IStop>('Stop', stopSchema);

File: shared/lib/models/stop.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'location.dart';

part 'stop.g.dart';

/// 버스 정류장 정보를 나타내는 클래스
@JsonSerializable()
class Stop extends Equatable {
  /// 정류장 고유 식별자
  @JsonKey(name: '_id')
  final String id;
  
  /// 정류장 이름
  final String name;
  
  /// 정류장 위치 정보
  final Location location;
  
  /// 이 정류장을 경유하는 노선 ID 목록
  final List<String> routes;
  
  /// 정류장 시설 정보 (대기실, 벤치 등)
  /// 예: {'hasShelter': true, 'hasBench': true}
  final Map<String, bool>? facilities;
  
  /// 정류장 설명
  final String? description;
  
  /// 정류장 주소
  final String? address;
  
  /// 정류장 이미지 URL
  final String? imageUrl;
  
  /// 추가 메타데이터
  final Map<String, dynamic>? metadata;

  const Stop({
    required this.id,
    required this.name,
    required this.location,
    required this.routes,
    this.facilities,
    this.description,
    this.address,
    this.imageUrl,
    this.metadata,
  });

  /// JSON에서 Stop 객체 생성
  factory Stop.fromJson(Map<String, dynamic> json) => _$StopFromJson(json);
  
  /// Stop 객체를 JSON으로 변환
  Map<String, dynamic> toJson() => _$StopToJson(this);

  @override
  List<Object?> get props => [
    id, 
    name, 
    location, 
    routes, 
    facilities,
    description, 
    address, 
    imageUrl, 
    metadata
  ];

  /// 새로운 속성으로 복사본 생성
  Stop copyWith({
    String? id,
    String? name,
    Location? location,
    List<String>? routes,
    Map<String, bool>? facilities,
    String? description,
    String? address,
    String? imageUrl,
    Map<String, dynamic>? metadata,
  }) {
    return Stop(
      id: id ?? this.id,
      name: name ?? this.name,
      location: location ?? this.location,
      routes: routes ?? this.routes,
      facilities: facilities ?? this.facilities,
      description: description ?? this.description,
      address: address ?? this.address,
      imageUrl: imageUrl ?? this.imageUrl,
      metadata: metadata ?? this.metadata,
    );
  }
}